From 843fd534ba87c2035ef54b14a85bd4c4839559c0 Mon Sep 17 00:00:00 2001
From: Brian Ma <chma0@nuvoton.com>
Date: Tue, 27 Jul 2021 10:29:36 +0800
Subject: [PATCH] Change Value property type to double and set scale default
 value

Based on the below spec, the Value property should be double:
https://github.com/openbmc/phosphor-dbus-interfaces/blob/master/yaml/xyz/openbmc_project/Sensor/Value.interface.yaml#L23
And the scale property is optional,
do not go to fail case when scale not set.

Signed-off-by: Brian Ma <chma0@nuvoton.com>
---
 dcmihandler.cpp | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/dcmihandler.cpp b/dcmihandler.cpp
index f8498f5..6c3a483 100644
--- a/dcmihandler.cpp
+++ b/dcmihandler.cpp
@@ -1007,6 +1007,7 @@ int64_t getPowerReading(sdbusplus::bus::bus& bus)
 
     // Return default value if failed to read from D-Bus object
     int64_t power = 0;
+    int64_t scale = 0;
     try
     {
         auto service = ipmi::getService(bus, SENSOR_VALUE_INTF, objectPath);
@@ -1014,8 +1015,17 @@ int64_t getPowerReading(sdbusplus::bus::bus& bus)
         // Read the sensor value and scale properties
         auto properties = ipmi::getAllDbusProperties(bus, service, objectPath,
                                                      SENSOR_VALUE_INTF);
-        auto value = std::get<int64_t>(properties[SENSOR_VALUE_PROP]);
-        auto scale = std::get<int64_t>(properties[SENSOR_SCALE_PROP]);
+        auto value = std::get<double>(properties[SENSOR_VALUE_PROP]);
+        // scale is not set in current node manager, handle it to avoid exeception
+        auto findIndex = properties.find(SENSOR_SCALE_PROP);
+        if (findIndex == properties.end())
+        {
+            log<level::ERR>("Cannot find property scale");
+        }
+        else
+        {
+            scale = std::get<int64_t>(properties[SENSOR_SCALE_PROP]);
+        }
 
         // Power reading needs to be scaled with the Scale value using the
         // formula Value * 10^Scale.
-- 
2.17.1

